### ISSUE TITLE
:scout crashes on slow filesystems

### ISSUE THREAD

**Issue by vifon** [ranger/ranger#278]
When I use `:scout` on a slow filesystem (it was `sshfs` but I doubt that it is specific to it) right after moving into a directory and start typing something, ranger crashes with the following stack trace:

```
ranger version: 1.6.1, executed with python 2.7.9
Locale: en_US.UTF-8
Current file: None
Traceback (most recent call last):
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/core/main.py", line 139, in main
    fm.loop()
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/core/fm.py", line 334, in loop
    ui.handle_input()
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/gui/ui.py", line 211, in handle_input
    self.handle_key(key)
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/gui/ui.py", line 145, in handle_key
    elif not DisplayableContainer.press(self, key):
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/gui/displayable.py", line 261,in press
    focused_obj.press(key)
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/gui/widgets/console.py", line 173, in press
    self.type_key(key)
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/gui/widgets/console.py", line 202, in type_key
    self.on_line_change()
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/gui/widgets/console.py", line 425, in on_line_change
    if cmd and cmd.quick():
  File "/home/vifon/local/lib/python2.7/site-packages/ranger/api/commands.py", line 400, inquick
    return self._make_cmd().quick()
  File "/home/vifon/.config/ranger/commands.py", line 1135, in quick
    if self._count(move=asyoutype) == 1 and self.AUTO_OPEN in self.flags:
  File "/home/vifon/.config/ranger/commands.py", line 1200, in _count
    deq = deque(cwd.files)
TypeError: 'NoneType' object is not iterable
```

I was able to reproduce it on two different computers in two radically different environments. I use the latest ranger master branch.

**Comment by hut** [comment]
Yes, `:scout` always crashes ranger when you use it before the directory is loaded. Although maybe only with certain options.

**Comment by saksham0808** [comment]
:search also crashes when the directory is not yet loaded. Noticeable when I access my music folder kept on a separate NTFS drive. Shouldn't a flag be set to prevent search from showing closest match dynamically while folder is loading?

**Comment by NoSuck** [comment]
I encountered this same bug for the first time today.  The traceback is more or less identical, except for `ranger version: 1.7.1, executed with python 3.4.3`.

**Comment by dabreese00** [comment]
Same her, in 1.7.0.  The crash seems to be reliable.  Steps to reproduce:
1. Open a large directory that takes a few seconds to scan
2. While loading, press "/" to open search prompt
3. While still loading, type something at search prompt
(4. I don't think I ever actually get to hit Enter)

Expected result:
Ranger waits till the directory is loaded, then executes any pre-typed searches.  Or, ranger just refuses to search while the directory is loading.

Actual result:
Ranger crashes at step 3.

Output (same as Vifon):
```
ranger version: 1.7.0, executed with python 2.7.10
Locale: en_US.UTF-8
Current file: None
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/ranger/core/main.py", line 139, in main
    fm.loop()
  File "/usr/lib64/python2.7/site-packages/ranger/core/fm.py", line 344, in loop
    ui.handle_input()
  File "/usr/lib64/python2.7/site-packages/ranger/gui/ui.py", line 211, in handle_input
    self.handle_key(key)
  File "/usr/lib64/python2.7/site-packages/ranger/gui/ui.py", line 145, in handle_key
    elif not DisplayableContainer.press(self, key):
  File "/usr/lib64/python2.7/site-packages/ranger/gui/displayable.py", line 261, in press
    focused_obj.press(key)
  File "/usr/lib64/python2.7/site-packages/ranger/gui/widgets/console.py", line 173, in press
    self.type_key(key)
  File "/usr/lib64/python2.7/site-packages/ranger/gui/widgets/console.py", line 202, in type_key
    self.on_line_change()
  File "/usr/lib64/python2.7/site-packages/ranger/gui/widgets/console.py", line 425, in on_line_change
    if cmd and cmd.quick():
  File "/usr/lib64/python2.7/site-packages/ranger/api/commands.py", line 397, in quick
    return self._make_cmd().quick()
  File "/usr/lib64/python2.7/site-packages/ranger/config/commands.py", line 1175, in quick
    if self._count(move=asyoutype) == 1 and self.AUTO_OPEN in self.flags:
  File "/usr/lib64/python2.7/site-packages/ranger/config/commands.py", line 1240, in _count
    deq = deque(cwd.files)
TypeError: 'NoneType' object is not iterable
```

**Comment by Midek** [comment]
same here, when moving with :travel over my secondary hard drive that takes a while to spin from sleep i got:
```
ranger version: 1.7.2, executed with python 3.5.2
Locale: pl_PL.UTF-8
Current file: None
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/ranger/core/main.py", line 140, in main
    fm.loop()
  File "/usr/lib/python3.5/site-packages/ranger/core/fm.py", line 344, in loop
    ui.handle_input()
  File "/usr/lib/python3.5/site-packages/ranger/gui/ui.py", line 211, in handle_input
    self.handle_key(key)
  File "/usr/lib/python3.5/site-packages/ranger/gui/ui.py", line 145, in handle_key
    elif not DisplayableContainer.press(self, key):
  File "/usr/lib/python3.5/site-packages/ranger/gui/displayable.py", line 261, in press
    focused_obj.press(key)
  File "/usr/lib/python3.5/site-packages/ranger/gui/widgets/console.py", line 173, in press
    self.type_key(key)
  File "/usr/lib/python3.5/site-packages/ranger/gui/widgets/console.py", line 202, in type_key
    self.on_line_change()
  File "/usr/lib/python3.5/site-packages/ranger/gui/widgets/console.py", line 425, in on_line_change
    if cmd and cmd.quick():
  File "/home/midek/.config/ranger/commands.py", line 1052, in quick
    if self._count(move=asyoutype) == 1 and self.AUTO_OPEN in self.flags:
  File "/home/midek/.config/ranger/commands.py", line 1117, in _count
    deq = deque(cwd.files)
TypeError: 'NoneType' object is not iterable
```

im on `Linux 4.8.6-1-ARCH #1 SMP PREEMPT Mon Oct 31 18:51:30 CET 2016 x86_64 GNU/Linux`

**Comment by Rom1deTroyes** [comment]
On Wednesday 07 Dec 2016 à 23:58:30 (-0800), Midek wrote:
> same here, when moving with :travel over my secondary hard drive that takes a while to spin from sleep i got:

Dicovering the :travel and :scout I try to ride over my ~/Télechargement
(Download folder on french system) : it always crashes (reproductible)
after typing :

:travel Té

No crash with :travel Vidéo (same 'é' char), nor with :scout Té

Hope it helps !

My version is older (debian jessie stable).

Linux 3.16.0-4-amd64 #1 SMP Debian 3.16.36-1+deb8u2 (2016-10-19) x86_64 GNU/Linux

ranger version: 1.6.1, executed with python 2.7.9
Locale: fr_FR.UTF-8
Current file: None
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ranger/core/main.py", line 134, in main
    fm.loop()
  File "/usr/lib/python2.7/dist-packages/ranger/core/fm.py", line 314, in loop
    ui.handle_input()
  File "/usr/lib/python2.7/dist-packages/ranger/gui/ui.py", line 193, in handle_input
    self.handle_keys(*keys)
  File "/usr/lib/python2.7/dist-packages/ranger/gui/ui.py", line 170, in handle_keys
    self.handle_key(key)
  File "/usr/lib/python2.7/dist-packages/ranger/gui/ui.py", line 142, in handle_key
    elif not DisplayableContainer.press(self, key):
  File "/usr/lib/python2.7/dist-packages/ranger/gui/displayable.py", line 261, in press
    focused_obj.press(key)
  File "/usr/lib/python2.7/dist-packages/ranger/gui/widgets/console.py", line 173, in press
    self.type_key(key)
  File "/usr/lib/python2.7/dist-packages/ranger/gui/widgets/console.py", line 202, in type_key
    self.on_line_change()
  File "/usr/lib/python2.7/dist-packages/ranger/gui/widgets/console.py", line 418, in on_line_change
    if cmd and cmd.quick():
  File "/usr/lib/python2.7/dist-packages/ranger/config/commands.py", line 1052, in quick
    if self._count(move=asyoutype) == 1 and self.AUTO_OPEN in self.flags:
  File "/usr/lib/python2.7/dist-packages/ranger/config/commands.py", line 1117, in _count
    deq = deque(cwd.files)
TypeError: 'NoneType' object is not iterable

++
    Romain
-- 
O toi qui sais ce que nous endurons ici
Ne nous oublie pas dans tes prières
	-+- Frank Herbert, Dune -+-

**Comment by roorco** [comment]
No news on this?

**Comment by FabianUntermoser** [comment]
Have the same issue using the latest version: 1.9.3.143.g5047d9e5-1. Any updates?

**Comment by toonn** [comment]
No one's looking into this yet. Would be a great way to contribute to ranger for anyone interested! We don't currently have any time to spare to look into this. Anyone should feel free to join us in #ranger at [Libera.Chat](https://libera.chat/) if they want to have a crack at this, especially if they're interested in becoming a maintainer.

**Comment by markus-bauer** [comment]
Is this really still an issue?
_count has this check:
https://github.com/ranger/ranger/blob/e3ca4e75daa82d6fcf8d12406adf8e187e4750a1/ranger/config/commands.py#L1706-L1707
So crashes because of files being None shouldn't happen.
 
I looked at the older versions that were mentioned here, like 1.7.2, and they don't have this check.
It seems this was added here:
https://github.com/ranger/ranger/commit/7f40b26887e299517781602a01e245895dbffd17

Can anyone post a traceback using a ranger version >= 1.8?
